src/pam_pivcard.8
.TH "PAM_PIVCARD" "8" "04/01/2015" "" ""
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.nh
.ad l
.SH "NAME"
pam_pivcard \- Module for PIV smartcard authentication
.SH "SYNOPSIS"
.HP \w'\fBpam_pivcard\&.so\fR\ 'u
\fBpam_pivcard\&.so\fR [\&.\&.\&.]
.SH "DESCRIPTION"
.PP
The \fBpam_pivcard\fR \s-1PAM\s0 module allows for system authentication using PIV and compatible smartcards (actually developed using a Yubikey). Additionally it can inject the user's plaintext password (stored encrypted on disk) for other \s-1PAM\s0 modules that require an actual authentication token. The latter requires a local (/etc/shadow) user account.
.PP
A helper binary, \fBpivhelper\fR, is provided which actually handles all smartcard processing. The helper is very simple and will proceed only if the effective user id is 0 (root). This way the calling process does not need to load the libraries required for smartcard authentication.
.SH "OPTIONS"
.PP
\fBtry_first_pass\fR
.RS 4
Use password already entered as PIN, if specified, otherwise inquire PIN.
.RE
.PP
\fBuse_first_pass\fR
.RS 4
Require an already entered password which is used as PIN.
.RE
.PP
\fBopensc_config=\fR\fB\fIpath\fR\fR
.RS 4
The opensc configuration file to use, if not specified the global configuration file (typically /etc/opensc.conf) is used.
.RE
.PP
\fBhelper=\fR\fB\fIpath\fR\fR
.RS 4
The helper utility that actually interworks with the smartcard by means of OpenSSL (engine), default is /usr/libexec/pivhelper.
.RE
.PP
\fBengine=\fR\fB\fIpath\fR\fR
.RS 4
The OpenSSL engine library to use, default is either /usr/lib64/engines/engine_pkcs11.so or /usr/lib/engines/engine_pkcs11.so depending on whether the system is 64 bit or 32 bit. Starting with engine_pkcs11-0.2.2 this seems to be changed to /usr/lib64/libpkcs11.so or /usr/lib/libpkcs11.so, respectively.
.RE
.PP
\fBpkcs11=\fR\fB\fIpath\fR\fR
.RS 4
The PKCS11 library to use, default is is either /usr/lib64/opensc-pkcs11.so or /usr/lib/opensc-pkcs11.so depending on whether the system is 64 bit or 32 bit.
.RE
.PP
\fBuserfile=\fR\fB\fIpath\fR\fR
.RS 4
The optional configuration file specifying users, processing modes and certificates. The format of this file is /etc/passwd style and consists of:
.TP
\fR\fB\fI<user>:<mode>:<comment>:<certificate>\fR\fR
.TP
.RS -4
\fR\fB\fI<user>\fR\fR is the same as in /etc/passwd.
\fR\fB\fI<mode>\fR\fR is the processing mode, a number in the range of 0-3.
\fR\fB\fI 0\fR\fR means that the public key stored on card is matched to the public key on disk and in case of a match challenge response authentication takes place.
\fR\fB\fI 1\fR\fR is like 0 with the addition that an on disk encrypted password file is decrypted and, if decryption was successful, the plaintext password is injected as the authentication token onto the PAM stack.
\fR\fB\fI 2\fR\fR means that the public key stored on card is matched to the public key on disk. In case of a match the disk encrypted password file is then decrypted and, if decryption was successful, the plaintext password is injected as the authentication token onto the PAM stack.
\fR\fB\fI 3\fR\fR is like 2 but without any public key matching, thus the specified on disk certificate does not need to exist.
\fR\fB\fI<certificate>\fR\fR is the pathname of the associated certificate which contains the public key related to the private key on the smartcard and \fR\fB\fI<comment>\fR\fR is an arbitary comment.
.RS 4
.RE
.PP
\fBpassfiledir=\fR\fB\fIpath\fR\fR
.RS 4
The directory the encrypted passwords are kept in, default is /etc/pam_pivcard/data.
.RE
.PP
\fBkeyfile=\fR\fB\fIreader:slot\fR\fR
.RS 4
This specifies the key slot on the smartcard used for authentication. For a PIV smartcard the following slots are possible:
.TP
.RS -4
\fR\fB\fI"01:01"\fR\fR key slot 9A (PIN required once), PIV Authentication
\fR\fB\fI"01:02"\fR\fR key slot 9C (PIN always required), Digital Signature
\fR\fB\fI"01:03"\fR\fR key slot 9D (PIN required once), for Key Management
\fR\fB\fI"01:04"\fR\fR key slot 9E (PIN never required), Card Authentication
The default is \fR\fB\fI"01:01"\fR\fR.
.PP
.RS 4
Note that the first two digits before the colon specify the reader slot of the card reader to be used. You are advised to use a private opensc.conf file with proper ignored_readers and allowed_readers (requires supplied patch) options to always have your PIV card available in slot 1 if possible.
.TP
The options \fBkeyfile2\fR through \fBkeyfile9\fR with similar semantics but an empty default exist. You can use them if you have e.g. a NFC reader and when inserting your token directly into a USB port makes your token appear as "02:xx".
.TP
Use the following command (adapted to your system) to list the readers as found by the pivhelper:
.TP
.RS -4
env OPENSC_CONF=<your opensc.conf> pkcs11-tool --module <path to opensc-pkcs11.so> --list-slots
.RS 4
.RE
.PP
\fBnopin\fR
.RS 4
Don't ask for a PIN, only usable with keyfile "xx:04" (requires supplied patch).
.RE
.PP
\fBnodevok\fR
.RS 4
Don't treat missing authentication token as an error when changing your password.
.RE
.PP
\fBdofail\fR
.RS 4
Return failure codes instead of ignore to the PAM stack when authenticating.
.SH "MODULE TYPES PROVIDED"
.PP
All module types (\fBaccount\fR, \fBauth\fR, \fBpassword\fR and \fBsession\fR) are provided, though \fBaccount\fR and \fBsession\fR currently do nothing.
.SH "RETURN VALUES"
.RE
.PP
PAM_SERVICE_ERR
.RS 4
Configuration error.
.RE
.PP
PAM_SUCCESS
.RS 4
Successful challenge response authentification (only if configuration file is used and only in mode 0 which is challenge response only), or password change possible or complete.
.RE
.PP
PAM_IGNORE
.RS 4
All other authentication cases if "dofail" is not specified or, if "dofail" is specified, in case of successful injection of the plaintext password onto the PAM stack.
.RE
.PP
PAM_AUTH_ERR
.RS 4
Authentication failure (e.g. missing or illegal PIN or unsuccessful card authentification or password file pathname length exceeded).
.RE
.PP
PAM_AUTHINFO_UNAVAIL
.RS 4
Problem starting the helper or a helper communication problem or e.g. an unacceptable user name.
.RE
.PP
PAM_PERM_DENIED
.RS 4
Password change not possible due to missing smartcard.
.RE
.PP
PAM_CRED_INSUFFICIENT
.RS 4
Missing or illegal smartcard PIN.
.RE
.PP
PAM_BUF_ERR
.RS 4
Out of memory.
.RE
.SH "SEE ALSO"
.PP
\fBopenssl\fR(1),
\fBopensc-tool\fR(1),
\fBpam.d\fR(5),
\fBpam\fR(8)
.SH "AUTHOR"
.PP
pam_pivcard was written by Andreas Steinmetz.


ChangeLog
1.0.0 Initial Release
1.0.1 make pivhelper self destruct via SIGALRM->SIGKILL after 10 seconds
1.0.2 add helpers to temporarily disable the pkcs11 engine


src/pam_pivcard.c
/*
 *
 * The pam_pivcard license version 1.0 This file is part of pam_pivcard,
 * the PIV (compatible) smartcard authentication module.
 * (C) 2015 Andreas Steinmetz, ast@domdv.de
 *
 * pam_pivcard is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License as published by the
 * Free Software Foundation; either version 2 or (at your option) any later
 * version.
 *
 * pam_pivcard is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License
 * for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with pam_pivcard; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
 *
 * As a special exception, if you compile this file and link it with
 * other works to produce a work based on this file, this file does not
 * by itself cause the resulting work to be covered by the GNU General
 * Public License. However the source code for this file must still be
 * made available in accordance with section (3) of the GNU General Public
 * License v2.
 *
 * This exception does not invalidate any other reasons why a work based
 * on this file might be covered by the GNU General Public License.
 *
 */

#define PAM_SM_AUTH
#define PAM_SM_ACCOUNT
#define PAM_SM_SESSION
#define PAM_SM_PASSWORD

#include <sys/socket.h>
#include <sys/stat.h>
#include <sys/wait.h>
#include <poll.h>
#include <fcntl.h>
#include <stdlib.h>
#include <unistd.h>
#include <string.h>
#include <stdio.h>
#include <security/pam_appl.h>
#include <security/pam_modules.h>
#include "common.h"

typedef struct
{
	int passvalid;
	int nopass;
	char pin[PINSIZE];
	char pass[PASSSIZE];
	char user[USERSIZE];
} MEM;

typedef struct
{
	char *item;
	int len;
	int index;
	int size;
} PARAM;

#define TRYFIRST	0
#define USEFIRST	1
#define NOPIN		2
#define NODEVOK		3
#define DOFAIL		4
#define OPENSC		5
#define HELPER		6
#define PASSDIR		7
#define ENGINE		8
#define PKCS11		9
#define CFGFILE		10
#define KEYFILE1	11
#define KEYFILE2	12
#define KEYFILE3	13
#define KEYFILE4	14
#define KEYFILE5	15
#define KEYFILE6	16
#define KEYFILE7	17
#define KEYFILE8	18
#define KEYFILE9	19
#define TOTALPARAMS	20

static PARAM config[TOTALPARAMS]=
{
	{"try_first_pass",0,TRYFIRST,0},
	{"use_first_pass",0,USEFIRST,0},
	{"nopin",0,NOPIN,0},
	{"nodevok",0,NODEVOK,0},
	{"dofail",0,DOFAIL,0},
	{"opensc_config=",14,OPENSC,CONFSIZE-12},
	{"helper=",7,HELPER,0},
	{"passfiledir=",12,PASSDIR,PASSSIZE},
	{"engine=",7,ENGINE,DIRSIZE},
	{"pkcs11=",7,PKCS11,DIRSIZE},
	{"userfile=",9,CFGFILE,CONFSIZE},
	{"keyfile=",8,KEYFILE1,PINSIZE},
	{"keyfile2=",9,KEYFILE2,PINSIZE},
	{"keyfile3=",9,KEYFILE3,PINSIZE},
	{"keyfile4=",9,KEYFILE4,PINSIZE},
	{"keyfile5=",9,KEYFILE5,PINSIZE},
	{"keyfile6=",9,KEYFILE6,PINSIZE},
	{"keyfile7=",9,KEYFILE7,PINSIZE},
	{"keyfile8=",9,KEYFILE8,PINSIZE},
	{"keyfile9=",9,KEYFILE9,PINSIZE},
};

static int starthelper(const char *helper)
{
	pid_t pid;
	int s[2];
	char bfr[16];
	struct stat stb;

	if(stat(helper,&stb)||!S_ISREG(stb.st_mode))return -1;

	if(socketpair(AF_UNIX,SOCK_STREAM|SOCK_CLOEXEC,0,s))return -1;
	sprintf(bfr,"%d",s[1]);

	switch((pid=fork()))
	{
	case 0: close(s[0]);
		unsetenv("LD_PRELOAD");
		unsetenv("LD_LIBRARY_PATH");
		switch(fork())
		{
		case 0:	fcntl(s[1],F_SETFD,fcntl(s[1],F_GETFD)&~FD_CLOEXEC);
			execl(helper,helper,bfr,NULL);
		case -1:exit(1);
		default:exit(0);
		}

	default:close(s[1]);
		waitpid(pid,NULL,0);
		return s[0];

	case -1:close(s[0]);
		close(s[1]);
		return -1;
	}
}

static int preprocess(pam_handle_t *pamh,const char **user,const char **pass,
	int try_first_pass,int use_first_pass,int nopin)
{
	int r;
	struct pam_conv *cnv;
	struct pam_message msg;
	struct pam_message *mptr=&msg;
	struct pam_response *rsp=NULL;

	if((r=pam_get_user(pamh,user,NULL))!=PAM_SUCCESS)return r;

	if(nopin)return PAM_SUCCESS;

	if(try_first_pass||use_first_pass)
	{
		if((r=pam_get_item(pamh,PAM_AUTHTOK,(const void **)pass))!=
			PAM_SUCCESS)return r;

		if(use_first_pass&&!*pass)return PAM_AUTH_ERR;
	}

	if(!*pass)
	{
		if((r=pam_get_item(pamh,PAM_CONV,(const void **)&cnv))!=
			PAM_SUCCESS)return r;

		msg.msg_style=PAM_PROMPT_ECHO_OFF;
		msg.msg="Enter PIN: ";

		if((r=cnv->conv(1,(const struct pam_message **)&mptr,&rsp,
			cnv->appdata_ptr))!=PAM_SUCCESS)return r;

		if(!(*pass=rsp->resp))return PAM_AUTH_ERR;
	}

	return PAM_SUCCESS;
}

static int notify(pam_handle_t *pamh,int flags,char *msg,int type)
{
	int r;
	struct pam_conv *cnv;
	struct pam_message pmsg;
	struct pam_message *mptr=&pmsg;
	struct pam_response *rsp=NULL;

	if(flags&PAM_SILENT)return PAM_SUCCESS;

	if((r=pam_get_item(pamh,PAM_CONV,(const void **)&cnv))!=PAM_SUCCESS)
		return r;

	pmsg.msg_style=type;
	pmsg.msg=msg;

	return cnv->conv(1,(const struct pam_message **)&mptr,
		&rsp,cnv->appdata_ptr);
}

static void clear(pam_handle_t *pamh,void *data,int error_status)
{
	MEM *mem=data;

	if(mem)
	{
		memclear(mem,0,sizeof(MEM));
		free(mem);
	}
}

static int parse(int argc,const char **argv,char **params)
{
	int i;
	int j;

	memset(params,0,TOTALPARAMS*sizeof(char *));

	for(j=0;j<argc;j++)
	{
		for(i=0;i<TOTALPARAMS;i++)if(!config[i].len)
		{
			if(!strcmp(argv[j],config[i].item))
			{
				params[i]=(char *)argv[j];
				break;
			}
		}
		else if(!strncmp(argv[j],config[i].item,config[i].len))
		{
			params[i]=(char *)argv[j]+config[i].len;
			break;
		}
	}

	if(!params[HELPER])params[HELPER]=DFTHELPER;
	if(!params[PASSDIR])params[PASSDIR]=DFTDIR;
	if(!params[KEYFILE1])params[KEYFILE1]=DFTFILE;
	if(!params[ENGINE])params[ENGINE]=DFTENGINE;
	if(!params[PKCS11])params[PKCS11]=DFTPKCS11;

	for(i=0;i<TOTALPARAMS;i++)if(params[i]&&config[i].size)
		if(strlen(params[i])>=config[i].size)return -1;

	return 0;
}

static int run(int cmd,int flag,char **params,const char *user,const char *pin,
	const char *pass,MESSAGE *msg)
{
	int r=-1;
	int fd;
	struct pollfd p;

	memset(msg,0,sizeof(MESSAGE));
	msg->cmdres=cmd;
	if(params[OPENSC])
	{
		strcpy(msg->openscconf,"OPENSC_CONF=");
		strcat(msg->openscconf,params[OPENSC]);
	}
	strcpy(msg->passfile,params[PASSDIR]);
	if(user)
	{
		strcat(msg->passfile,"/");
		strcat(msg->passfile,user);
	}
	strcpy(msg->engine,params[ENGINE]);
	strcpy(msg->pkcs11,params[PKCS11]);
	if(params[CFGFILE])strcpy(msg->cfgfile,params[CFGFILE]);
	for(fd=0;fd<MAXFILES;fd++)strcpy(msg->keyfile[fd],params[KEYFILE1+fd]);
	if(pin)strcpy(msg->pin,params[NOPIN]?"":pin);
	if(pass)strcpy(msg->pass,pass);
	msg->nopin=params[NOPIN]?1:0;
	msg->precheck=flag;

	if((fd=starthelper(params[HELPER]))==-1)goto err1;
	if(write(fd,msg,sizeof(MESSAGE))!=sizeof(MESSAGE))goto err2;
	p.events=POLLIN;
	p.fd=fd;
	if(poll(&p,1,10000)!=1||!(p.revents&POLLIN))goto err2;
	if(read(fd,msg,sizeof(MESSAGE))!=sizeof(MESSAGE))goto err2;
	r=0;
err2:	close(fd);
err1:	if(r)memclear(msg,0,sizeof(MESSAGE));
	return r;
}

static int pinchk(const char *pin)
{
	int i;

	if(pin)
	{
		for(i=0;pin[i];i++)if(pin[i]<'0'||pin[i]>'9')
			return PAM_AUTH_ERR;
		if(i>=PINSIZE)return PAM_AUTH_ERR;
	}
	return PAM_SUCCESS;
}

static int puchk(const char *user,char **params)
{
	if(*params[PASSDIR]!='/')return PAM_SERVICE_ERR;
	if(strlen(params[PASSDIR])+strlen(user)+1>=DIRSIZE)return PAM_AUTH_ERR;
	if(!*user||!strcmp(user,".")||!strcmp(user,"..")||strchr(user,'/')||
		strlen(user)>=USERSIZE)return PAM_AUTHINFO_UNAVAIL;
	return PAM_SUCCESS;
}

static int ignore(pam_handle_t *pamh,int flags,char **params,MESSAGE *msg,
	char *txt)
{
	int r=PAM_SUCCESS;

	switch(msg->cmdres)
	{
	case NOCARD:
		if(params[NODEVOK])break;
	case ENGFAIL:
	case FILEFAIL:
	case CRYPTOFAIL:
		r=PAM_PERM_DENIED;
	case NOUSER:
		break;
	}

	if(msg->cmdres&&r==PAM_SUCCESS)notify(pamh,flags,txt,PAM_TEXT_INFO);

	memclear(msg,0,sizeof(MESSAGE));
	return r;
}

PAM_EXTERN int pam_sm_authenticate(pam_handle_t *pamh,int flags,int argc,
	const char **argv)
{
	int r;
	const char *user=NULL;
	const char *pass=NULL;
	char *params[TOTALPARAMS];
	MESSAGE msg;

	if(parse(argc,argv,params))return PAM_SERVICE_ERR;

	if((r=preprocess(pamh,&user,&pass,params[TRYFIRST]?1:0,
		params[USEFIRST]?1:0,params[NOPIN]?1:0)!=PAM_SUCCESS))goto err1;

	if((r=pinchk(pass))!=PAM_SUCCESS)goto err1;
	if((r=puchk(user,params))!=PAM_SUCCESS)goto err1;

	if(run(GETPASS,0,params,user,pass,NULL,&msg))
	{
		r=PAM_AUTHINFO_UNAVAIL;
		goto err1;
	}
	if(msg.cmdres!=OK)r=PAM_AUTH_ERR;
	else if(msg.nopass)r=PAM_SUCCESS;
	else
	{
		r=PAM_IGNORE;
		pam_set_item(pamh,PAM_AUTHTOK,(const void **)(&msg.pass));
	}

	memclear(&msg,0,sizeof(msg));
err1:	if(r!=PAM_SUCCESS&&!params[DOFAIL])r=PAM_IGNORE;
	return r;
}

PAM_EXTERN int pam_sm_chauthtok(pam_handle_t *pamh,int flags,int argc,
	const char **argv)
{
	int r;
	int nopass=0;
	const char *user=NULL;
	const char *pass=NULL;
	const char *newpass=NULL;
	char *params[TOTALPARAMS];
	MEM *mem;
	MESSAGE msg;

	if(parse(argc,argv,params))return PAM_SERVICE_ERR;

	if(flags&PAM_PRELIM_CHECK)
	{
		if((r=pam_get_item(pamh,PAM_AUTHTOK,(const void **)&newpass))!=
			PAM_SUCCESS)if(r!=PAM_PERM_DENIED)return r;

		if(newpass&&strlen(newpass)>=sizeof(msg.pass))
			return PAM_AUTH_ERR;

		if((r=preprocess(pamh,&user,&pass,0,0,params[NOPIN]?1:0))!=
			PAM_SUCCESS)return r;

		if(pinchk(pass)!=PAM_SUCCESS)return PAM_CRED_INSUFFICIENT;
		if((r=puchk(user,params))!=PAM_SUCCESS)return r;

		if(run(SETPASS,1,params,user,pass,NULL,&msg))
			return PAM_AUTHINFO_UNAVAIL;
		nopass=msg.nopass;

		if(msg.cmdres!=OK)
		{
			if((r=ignore(pamh,flags,params,&msg,"Warning: token "
				"not available"))!=PAM_SUCCESS)return r;
		}
		else memclear(&msg,0,sizeof(msg));

		if(!(mem=malloc(sizeof(MEM))))return PAM_BUF_ERR;

		strcpy(mem->pin,params[NOPIN]?"":pass);
		strcpy(mem->user,user);
		if(newpass)
		{
			strcpy(mem->pass,newpass);
			mem->passvalid=1;
		}
		else mem->passvalid=0;
		mem->nopass=nopass;

		if((r=pam_set_data(pamh,"pam_pivcard",mem,clear))!=PAM_SUCCESS)
		{
			memclear(mem,0,sizeof(MEM));
			free(mem);
			return r;
		}

		return PAM_SUCCESS;
	}

	if(!(flags&PAM_UPDATE_AUTHTOK))return PAM_SERVICE_ERR;

	if((r=pam_get_data(pamh,"pam_pivcard",(const void **)&mem))!=
		PAM_SUCCESS)
	{
		if(r!=PAM_NO_MODULE_DATA)return r;

		if((r=pam_get_item(pamh,PAM_AUTHTOK,(const void **)newpass))!=
			PAM_SUCCESS)return r;

		if(!newpass||strlen(newpass)>=sizeof(msg.pass))
			return PAM_AUTH_ERR;

		if((r=preprocess(pamh,&user,&pass,0,0,params[NOPIN]?1:0))!=
			PAM_SUCCESS)return r;

		if(pinchk(pass)!=PAM_SUCCESS)return PAM_CRED_INSUFFICIENT;
	}
	else if(mem->nopass)return PAM_SUCCESS;
	else if(!mem->passvalid)
	{
		if((r=pam_get_item(pamh,PAM_AUTHTOK,(const void **)&newpass))!=
			PAM_SUCCESS||!newpass)return r;

		if(strlen(newpass)>=sizeof(msg.pass))return PAM_AUTH_ERR;
		strcpy(mem->pass,newpass);
	}

	if(run(SETPASS,0,params,mem?mem->user:user,
		params[NOPIN]?"":mem?mem->pin:pass,mem?mem->pass:newpass,&msg))
		return PAM_AUTHINFO_UNAVAIL;

	if(msg.cmdres!=OK)
	{
		if((r=ignore(pamh,flags,params,&msg,"Warning: token password "
			"not updated"))!=PAM_SUCCESS)return r;
	}
	else memclear(&msg,0,sizeof(msg));

	return PAM_SUCCESS;
}

PAM_EXTERN int pam_sm_open_session(pam_handle_t *pamh,int flags,int argc,
	const char **argv)
{
	return PAM_IGNORE;
}

PAM_EXTERN int pam_sm_close_session(pam_handle_t *pamh,int flags,int argc,
	const char **argv)
{
	return PAM_IGNORE;
}

PAM_EXTERN int pam_sm_acct_mgmt(pam_handle_t *pamh,int flags,int argc,
	const char **argv)
{
	return PAM_IGNORE;
}

PAM_EXTERN int pam_sm_setcred(pam_handle_t *pamh,int flags,int argc,
	const char **argv)
{
	return PAM_IGNORE;
}

#ifdef PAM_STATIC

struct pam_module _pam_neototp_modstruct=
{
	"pam_pivcard",
	pam_sm_authenticate,
	pam_sm_setcred,
	pam_sm_acct_mgmt,
	pam_sm_open_session,
	pam_sm_close_session,
	pam_sm_chauthtok
};

#endif
